fix(packet_transport): Add initialization for sensor configuration in packet transport

TMaYaD opened the pull request:
Ensure that the sensor configuration list is created if it does not exist when validating packet transport sensors.

Without this initialisation, reloading config after `CORE.reset()` throws missing key errors.

# What does this implement/fix?

This reinitialisation is required for reloading the config after `CORE.reset()` as done here https://github.com/TMaYaD/esphome-plus/blob/main/esphome_plus/two_stage.py#L54

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

TMaYaD committed:
fix(packet_transport): Add initialization for sensor configuration in packet transport

Ensure that the sensor configuration list is created if it does not exist when validating packet transport sensors.

Without this initialisation, reloading config after `CORE.reset()` throws missing key errors.

TMaYaD commented:
Here is a more detailed description of the bug generated using AI:

The problem is on line 125:
```python
# Build a list of sensors for this platform
CORE.data[DOMAIN] = {CONF_SENSORS: []}
```
This line is executed at module import time (when the module is first imported), and it creates a global data structure in `CORE.data[DOMAIN]`. However, this is not reset when CORE.reset() is called!
Looking at the `CORE.reset()` method, I can see that it resets `self.data = {}`, but the issue is that the initialisation is not run again due to how python modules work.
The Root Cause:
When you load the config first, the packet_transport module gets imported
Line 125 executes: `CORE.data[DOMAIN] = {CONF_SENSORS: []}`
Later, when `validate_packet_transport_sensor()` is called (line 140), it appends to this list
When you call `run_esphome()`, `CORE.reset()` clears `CORE.data = {}`
But when the module is imported again, the already loaded copy is used and line 125 does not runs again.
The Fix:
The issue is that CORE.data[DOMAIN] is being initialized at module import time, but it should be initialized when needed.


I'm being conservative here in not removing the original initialisation at module level. But I have tested it to run successfully without it.

clydebarrow commented:
`CORE.reset()` is only there for use in tests and isn't called in normal operation. I am a little reluctant to add code that is solely to fix an issue in a 3rd party project that is using the ESPHome code in unintended ways. 

That said, if the fix is cleaned up by removing the static (import) initialisation, and using a more idiomatic lazy-initialisation approach with `.setdefault()` and `.get()` then it becomes a code quality improvement which I'd have no qualms about.

TMaYaD committed:
refactor(packet_transport): Simplify sensor configuration initialization

Streamline the initialization of the sensor configuration list by using setdefault to ensure it exists before appending new sensors. This change prevents potential key errors when accessing the sensor list.

TMaYaD commented:
Thank you for the suggestion. This is good to go I hope.
